torch/export/unflatten.py

```diff
@@ -584,7 +579,74 @@ def unflatten(
     return UnflattenedModule(module, flat_args_adapter)
 
 
-def _inplace_buffer_mutations(graph: torch.fx.Graph, graph_signature) -> None:
+def _collect_intermediate_values_of_buffer_mutations(
+    graph: torch.fx.Graph,
+    graph_signature: ExportGraphSignature,
+) -> Dict[str, List[torch.fx.Node]]:
+    """
+    Collect the intermediate values of buffer mutations in a graph,
+    along with the module call fqns that create and use them. Later,
+    in each fqn associated with an intermediate value we will install
+    a corresponding attribute, so that it can be updated and read.
+
+    Example: in the following graph, suppose that buf_in and buf_out
+    are the input and output values of a buffer.
+
+        buf_in = placeholder()
+        ...
+        ival1 = f0(buf_in, ...)  # inside self.n0(...)
+        ...
+        ival2 = f1(ival1, ...)  # inside self.n1(...)
+        ...
+        buf_out = f2(ival2, ...)  # inside self.n2(...)
+        return buf_out, ...
+
+    Here ival1 and ival2 are intermediate values created inside
+    calls to n0 and n1 respectively, and used inside calls to
+    n1 and n2 respectively. Thus our analysis will produce
+    {n0: [ival1], n1: [ival1, ival2], n2: [ival2]}.
+    """
+    output_node = next(iter(reversed(graph.nodes)))
+    assert output_node.op == "output" and len(output_node.args) == 1
+    return_args = output_node.args[0]
+
+    mutation_node_to_buffer = graph_signature.buffers_to_mutate
```

Comment:
Hmm this won't work for training IR cause because we don't have this information...

Reply:
And fixing this is probably complicated, maybe we should document it in the unflatten function, could you file a task to track this?

Reply:
Note that this is only needed when buffer updates have been functionalized away. If the code contains direct buffer updates, then none of this is required. So it should be fine.

Reply:
Hmm, I was wrong. It turns out that the output node, rather than the input node, of mutation is what is threaded through the rest of the program, so every buffer mutation won't have the placeholder node corresponding to the buffer.

I think temporarily decomposing the training IR is overkill. How can I detect mutating ops?